networking/p2p/src/test/java/tech/pegasys/artemis/networking/p2p/discovery/discv5/NodeRecordConverterTest.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.networking.p2p.discovery.discv5;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static tech.pegasys.artemis.networking.p2p.discovery.discv5.NodeRecordConverter.convertToDiscoveryPeer;
+
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Optional;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.units.bigints.UInt64;
+import org.ethereum.beacon.discovery.schema.EnrField;
+import org.ethereum.beacon.discovery.schema.IdentitySchema;
+import org.ethereum.beacon.discovery.schema.NodeRecord;
+import org.ethereum.beacon.discovery.schema.NodeRecordFactory;
+import org.junit.jupiter.api.Test;
+import tech.pegasys.artemis.networking.p2p.discovery.DiscoveryNodeId;
+import tech.pegasys.artemis.networking.p2p.discovery.DiscoveryPeer;
+
+class NodeRecordConverterTest {
+
+  private static final DiscoveryNodeId NODE_ID =
+      new DiscoveryNodeId(
+          Bytes.fromHexString(
+              "0x833246C198388F1B5E06EF1950B0A6705FBF6370E002656CDA2C6C803C06258D"));
+  private static final Bytes PUB_KEY =
+      Bytes.fromHexString("0x0295A5A50F083697FF8557F3C6FE0CDF8E8EC2141D15F19A5A45571ED9C38CE181");
+  private static final Bytes IPV6_LOCALHOST =
+      Bytes.fromHexString("0x00000000000000000000000000000001");
+
+  @Test
+  public void shouldNotConvertRecordWithNoIp() {
+    assertThat(convertNodeRecordWithFields()).isEmpty();
+  }
+
+  @Test
+  public void shouldNotConvertRecordWithIpButNoPort() {
+    assertThat(
+            convertNodeRecordWithFields(
+                new EnrField(EnrField.IP_V4, Bytes.wrap(new byte[] {127, 0, 0, 1}))))
+        .isEmpty();
+  }
+
+  @Test
+  public void shouldNotConvertRecordWithIpAndUdpPortButNoTcpPort() {
+    assertThat(
+            convertNodeRecordWithFields(
+                new EnrField(EnrField.IP_V4, Bytes.wrap(new byte[] {127, 0, 0, 1})),
+                new EnrField(EnrField.UDP_V4, 30303)))
+        .isEmpty();
+  }
+
+  @Test
+  public void shouldUseV4IpIfV6PortSpecifiedWithNoV6Ip() {
+    assertThat(
+            convertNodeRecordWithFields(
+                new EnrField(EnrField.IP_V4, IPV6_LOCALHOST), new EnrField(EnrField.TCP_V6, 30303)))
```

Comment:
Is this valid - should we catch this?  If an IP_V4 field contains and IPV6 address?

Reply:
This is a result of me having the fallback from IP_V6 to IP_V4 rather than for port.  The test has been changed around to match the now-fixed code.

Although we would still work if there was an IPV6 address in a V4 field - I think I'm ok with that.  Maybe the discovery library should validate that the value it gets is actually the right number of bytes but I think it works out ok regardless - the data was signed by the node so it hasn't been tampered with and we may wind up connecting to a node that others can't because we detect the version rather than insisting it match the field.